### ISSUE TITLE
General event support

### ISSUE THREAD

**Issue by jaredjj3** [opensheetmusicdisplay/opensheetmusicdisplay#1056]
##  Proposal

I think that OSMD integration/adoption could be enhanced by creating a general event (not just DOM) system that callers can hook into.

Hopefully these contrived examples are illustrative of the possibilities:

```ts
type OSMDEvent<P, M = any> = { id: number; metadata: M; payload: P };

type CursorChangeEvent = OSMDEvent<{
  cursor: Cursor;
}>;

enum GraphicType {
  Note,
  Lyric,
}

type GraphicEvent = OSMDEvent<
  | { type: GraphicType.Note; note: Note }
  | { type: GraphicType.Lyric; lyric: Lyric, syllabic: boolean, text: string }
>;

// Non-DOM event: Vibrate a device every time the cursor changes.
osmd.on("cursorchange", (event: CursorChangeEvent) => {
  makeHapticFeedback(HapticFeedbackIntensity.Gentle, Duration.ms(10));
});

// Color a note head whenever it's hovered.
osmd.on("graphichover", (event: GraphicEvent) => {
  if (event.payload.type === GraphicType.Note) {
    const { note } = event.payload;
    if (noteColorer.currentlyColoredNote && noteColorer.currentlyColoredNote !== note) {
      noteColorer.restoreOriginalColor(noteColorer.currentlyColoredNote);
    }
    noteColorer.color(note, 'osmd-orange');
  }
});

// Announce the entire word of a lyric when it's clicked.
osmd.on("graphicclick", (event: GraphicEvent) => {
  if (event.payload.type === GraphicType.Lyric) {
    const { lyric, syllabic, text } = event.payload;
    const word = syllabic ? [text, ...getRemainingSyllablesAfter(lyric)].map(withoutTrailingHyphens).join('') : text;
    if (screenreader.canPronounce(word)) {
      screenreader.announce(word);
    }
  }
```

## Discussion

I think this can be done in a way that is not invasive and can be introduced gradually.

A possible solution could be to create an event bus that allows OSMD callers to hook into. Internally, the event bus would maintain a mapping of event names to arrays of listeners. Within the OSMD codebase, this event bus would be used to dispatch events with payloads that correspond to the event name.

One of the drawbacks of this approach is that it is the caller's responsibility for cleaning up listeners, since OSMD would not know nor care about the context outside of its purview. Perhaps there's something clever that can be done with [WeakRef](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef), but I'm unsure.

Another benefit of this approach is that it may surface refactoring opportunities within the OSMD codebase. For example, if there's a `cursorchange` being dispatched from several methods, it _might_ suggest that there's a common behavior that can be abstracted out amongst them. It also might suggest that the event dispatching is done in the wrong place as well as a bunch of other possibilities.

## Related Issues

#262
0xfe/vexflow#371

**Comment by jaredjj3** [comment]
I'm going to be offline for most of the weekend, but I'd like to share what I've come up with so far.

I created an [EventBus](https://github.com/jaredjj3/stringsync/blob/22d74e5b0cb97ed6813d4854088b9a902291b914/web/src/lib/EventBus/EventBus.ts) utility whose purpose is to allow callers to dispatch, subscribe, and unsubscribe events. This is the glue between OSMD and the rest of my application. The map of events to payloads is defined [here](https://github.com/jaredjj3/stringsync/blob/22d74e5b0cb97ed6813d4854088b9a902291b914/web/src/lib/MusicDisplay/types.ts#L6).

Next, I created an [SVGEventProxy](https://github.com/jaredjj3/stringsync/blob/22d74e5b0cb97ed6813d4854088b9a902291b914/web/src/lib/MusicDisplay/SVGEventProxy.ts#L16) whose purpose is to capture low level events, massage the data, and then use the `eventBus` to dispatch new events with the data I want. I set it up such that the caller (i.e. `OpenSheetMusicDisplay`) can specify a subset of events to listen to. This would ultimately be configurable by the user in the OSMD options, but I haven't ironed out those details. 

FWIW, each `OpenSheetMusicDisplay` instance will only have 1 `SVGEventProxy`.  By extension, each instance will only 1 set of event listeners.

An example of this in action is in the [onClick](https://github.com/jaredjj3/stringsync/blob/22d74e5b0cb97ed6813d4854088b9a902291b914/web/src/lib/MusicDisplay/SVGEventProxy.ts#L81) handler. I'm skeptical on how reliable [Element.closest](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest) is, but it seems to work ok. I'm also wary that this approach will break if the class names on the noteheads change.

This is diverges from [the solution I mentioned in the other comment](https://jsfiddle.net/gristow/j0j14sos/). After studying, I realized that event handlers were being installed on the SVG element as well as the all of elements that constitute each note. I'm going to try to do everything from the SVG element handler only, but I sense that I might have to do something similar.

**Comment by jaredjj3** [comment]
I'd like to make a progress update. I was able to make a working version of the event system, but I had to leverage the concept of iterator snapshots described in https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/139#issuecomment-903201412. Anyone that wants to adopt this event system directly would also need that use that solution as well.

The core implementation is contained within the [SVGEventProxy](https://github.com/jaredjj3/stringsync/blob/507fd3c96266e9d74a3931422df1c2f70ad865e9/web/src/lib/MusicDisplay/SVGEventProxy.ts) (link is pinned at rough solution commit).

One of my goals before trying to "publish" this as a community example is to make the solution reasonably maintainable. In the code's current state, it is already becoming difficult to conditionally dispatch certain events. For example, if the user is dragging a cursor, we don't want to dispatch `selectionupdated` events. Right now, I'm imperatively juggling state within the `SVGEventProxy` and I can see this easily ballooning into something unruly. My plan is to create state machine(s) that would allow me to add more events in a declarative manner. I think this will solve the maintainability issue. I plan to use the [xstate](https://xstate.js.org/) library.

Another goal is to make this solution resource efficient. I think I've generally accomplished this (throttling `mousemove` and `touchmove` events, installing 1 click handler per (event type, svg element), etc.), but I think there are micro optimizations that I can still make. For example, when determining if a cursor is being "hit" (on hovered, mousedown, touchdown, etc.), I suspect that calling `getBoundingClientRect` may be expensive in some contexts. To optimize, I could cache the hit and use it to shortcut the next calculation. All in all, I want to limit querying the DOM for events that may be dispatched frequently.

Here is a demo of me dispatching `mousemove`, `cursordragstarted`, and `cursordragended` events:

https://user-images.githubusercontent.com/19232300/131683654-1706861a-84c3-43e4-8548-429906b12be8.mov

**Comment by conde2** [comment]
This is amazing, I would love to see this in OSMD ! The sponsor build has some web interaction manager already coded, maybe you can talk to the OSMD team about it and make a integrated solution.

**Comment by jaredjj3** [comment]
# General Event Design Proposal

## Overview

I'm considering my example finished. The solution met all of my goals.

- The set of events that are listened to is configurable.
- It does not install listeners on _every_ element of interest. It only installs listeners on the `<svg>` element.
- It runs efficiently. Let `n` be the number of cursor positions in an `osmd` instance. The worst-case runtime of a music display lookup by position is `O(log(n))`.
- It throttles `mousemove` and `touchmove` events to limit unnecessary power consumption and laggy UX.
- It is maintainable. The pointer state machine is easy to unit test, suggesting that the code is abstracted reasonably. The application code does not probe state, implying that the pointer state machine is well encapsulated.

I'm still actively developing the solution, but I anticipate that the high-level approach will remain unchanged. I want to share what I've found.

**I do not intend to prescribe an exact implementation here or in the future unless someone requests it.** The evolving code is publicly available in the [stringsync repository](https://github.com/jaredjj3/stringsync). I've included some implementations in line with the abstract code.

## Architecture

 ### Glossary
 
- A `pointer` is a mouse cursor or finger touch.
- A `pointerTarget` describes OSMD object(s) under a pointer.
 - `<svg>` refers to the raw SVG element where OSMD renders. It is the mechanism that captures user actions by emitting low-level DOM events.
 - `SVGEventProxy` is a utility that orchestrates translating low-level DOM events into `pointerTarget`s.  This class is similar to `SVGInteraction` ([fiddle](https://jsfiddle.net/gristow/j0j14sos/) by @gristow). The difference is that it does not maintain pointer state.
 - `MusicDisplayLocator` is a service that locates OSMD object(s) temporally and positionally.
 - `PointerStateMachine` is a state machine for the `pointer`.
 - `KeepAlive` is a service that continuously calculates whether or not the `pointer`'s hover state has changed. This service can be omitted entirely or replaced with a service that knows when something movable _and_ interactable has moved. I plan to keep this around while I define what objects are movable and interactable.
 - `EventBus` dispatches arbitrary events to subscribers.
 - `Application` refers to my OSMD use case.

### Diagram

![Screen Shot 2021-09-04 at 2 13 31 PM](https://user-images.githubusercontent.com/19232300/132104591-1ea27c33-a845-4782-b6ec-07213c01a88b.png)
 
 ## Implementation
 
Imagine we want to do something whenever a user drags an OSMD object. We want to do something like this:

```ts
(async () => {
  const osmd = new OpenSheetMusicDisplay('osmd-container');
  await osmd.load('foo.xml');

  const svg = osmd.Drawer.Backends[0].getSvgElement();
  const eventBus = new EventBus();
  const locator = new MusicDisplayLocator();
  const pointerStateMachine = new PointerStateMachine('idle', eventBus);
  const svgEventProxy = new SVGEventProxy(svg, osmd, locator, pointerStateMachine);

  svgEventProxy.install(['mousedown', 'mousemove']);

  eventBus.subscribe('drag', (payload) => {
    console.log(`the dragged object is: ${payload.draggedObject}`);
  });
})();
```
 
 ### EventBus
 
The `EventBus` is straightforward to implement. It attaches subscribers to an event name. When dispatching an event, it synchronously invokes those subscribers with a payload.

```ts
class EventBus {
  subscribe<P>(eventName: string, callback: (payload: P) => void) {}
  dispatch<T, P>(eventName: T, payload: P) {}
}
```

[Example EventBus implementation](https://github.com/jaredjj3/stringsync/blob/8ba53165956197ecd2782465301e828876a0089d/web/src/lib/EventBus/EventBus.ts)

### MusicDisplayLocator

The `MusicDisplayLocator` is a performance-sensitive service because the `SVGEventProxy` calls it every time the pointer moves. It _must_ quickly find OSMD objects that contain a given position. A performant `MusicDisplayLocator` will use less device power and prevent the user from perceiving lag.

NB: I consciously chose to store OSMD object `hits` in an array because it's possible for many objects to contain a single point. It is ultimately up to the caller to decide how/if to calculate the significance of each hit.

```ts
type LocateResult = {
  timeMs: number;
  position: Position;
  hits: any[];
};

class MusicDisplayLocator {
  locateByTimeMs(timeMs: number): LocateResult {}
  locateByPosition(position: Position): LocateResult {}
}
```

[Example MusicDisplayLocator implementation](https://github.com/jaredjj3/stringsync/blob/8ba53165956197ecd2782465301e828876a0089d/web/src/lib/MusicDisplay/MusicDisplayLocator.ts)

### PointerStateMachine

The pointer state machine takes `LocateResult`s in the form of events and _applies_ them to produce a new state and context for the `pointer`. The event type matters even if the payloads are the same because state changes should depend on it. For example, you expect a click event to yield different results from a move event.

I think making a proper state machine that handles synchronous and asynchronous events is very difficult. I used the [xstate](https://xstate.js.org/) library, but I provide an example for illustrative purposes.

```ts
type PointerEvent = {
  type: string;
  payload: LocateResult;
}

class PointerStateMachine<Ctx = any> {
  ctx: Ctx = {} as any;
  state: string;
  eventBus: EventBus;

  constructor(initialState: string, eventBus: EventBus) {
    this.state = initialState;
    this.eventBus = eventBus;
  }

  send(event: PointerEvent) {
    const nextState = this.getNextState(this.ctx, this.state, event);
    const nextCtx = this.getNextCtx(this.ctx, this.state, event);
    const prevState = this.state;
    this.atomicallyUpdate(nextCtx, nextState);
    this.onTransition(nextCtx, prevState, nextState);
  }

  private getNextState(ctx: Ctx, state: string, event: PointerEvent): string {
    throw new NotImplementedError();
  }

  private getNextCtx(ctx: Ctx, state: string, event: PointerEvent): Ctx {
    throw new NotImplementedError();
  }

  private onTransition(ctx: Ctx, state: string, nextState: string) {
    if (state === 'selecting' && nextState === 'moving') {
      this.eventBus.dispatch('drag', { draggedObject: ctx.selectedObject });
    }
  }

  private atomicallyUpdate(ctx: Ctx, state: string) {
    this.ctx = ctx;
    this.state = state;
  }
}
```

[Example PointerStateMachine implementation](https://github.com/jaredjj3/stringsync/blob/8ba53165956197ecd2782465301e828876a0089d/web/src/lib/MusicDisplay/pointerMachine.ts)

### SVGEventProxy

The `SVGEventProxy` is the orchestrator of this event system. It decides when to locate OSMD objects and how to pipe `LocateResult`s into events that the `PointerStateMachine` can understand. It prevents the need to work with low-level DOM objects outside of the class.

Callers should not try to hook into the `SVGEventProxy` directly. Instead, events are piped into _some_ state machine and consumed indirectly via the `EventBus`. By maintaining this, callers remain decoupled from the `SVGEventProxy`, and they enjoy the benefits of a managed lifecycle from the state machine.

```ts
class SVGEventProxy {
  svg: SVGElement;
  osmd: OpenSheetMusicDisplay;
  locator: MusicDisplayLocator;
  pointerStateMachine: PointerStateMachine;

  constructor(
    svg: SVGElement,
    osmd: OpenSheetMusicDisplay,
    locator: MusicDisplayLocator,
    pointerStateMachine: PointerStateMachine
  ) {
    this.svg = svg;
    this.osmd = osmd;
    this.locator = locator;
    this.pointerStateMachine = pointerStateMachine;
  }

  install(eventNames: Array<keyof SVGElementEventMap>) {
    for (const eventName of eventNames) {
      if (eventName === 'mousedown') {
        this.svg.addEventListener(eventName, this.onMouseDown.bind(this));
      } else if (eventName === 'mousemove') {
        this.svg.addEventListener(eventName, this.onMouseMove.bind(this));
      } else {
        throw new NotImplementedError();
      }
    }
  }

  private onMouseDown(event: SVGElementEventMap['mousedown']) {
    const { x, y } = this.getSVGPosition(event);
    const locateResult = this.locator.locateByPosition(x, y);
    this.pointerStateMachine.send({ type: 'pointerdown', locateResult });
  }

  private onMouseMove(event: SVGElementEventMap['mousemove']) {
    const { x, y } = this.getSVGPosition(event);
    const locateResult = this.locator.locateByPosition(x, y);
    this.pointerStateMachine.send({ type: 'pointermove', locateResult });
  }

  private getSVGPosition(position: ClientPosition): Position {
    if (!this.osmd.GraphicSheet) {
      return { x: 0, y: 0 };
    }
    const domPoint = new PointF2D(position.clientX, position.clientY);
    return this.osmd.GraphicSheet.domToSvg(domPoint);
  }
}
```

[Example SVGEventProxy implementation](https://github.com/jaredjj3/stringsync/blob/8ba53165956197ecd2782465301e828876a0089d/web/src/lib/MusicDisplay/SVGEventProxy.ts)

## Demo

I sometimes hold down the mouse, tap it, or idle it in this video, but it's impossible to tell when I'm doing that without the console. Try to focus on the console logs and assume the system is doing the "right thing."

https://user-images.githubusercontent.com/19232300/132107564-4fa88363-99cd-48eb-9c46-993925aeb999.mov

**Comment by conde2** [comment]
That looks amazing! Hope this gets integrated with OSMD. You should colaborate in the sponsor tier, this could be easyly used togheter with the audio player =)

OSMD team need to looks this asap, great addition.

**Comment by trickart73** [comment]
Hey @jaredjj3 ! First of all, thanks so much for your work for improving OSMD !
What you did looks really amazing and I'm trying to implement the same kind of stuff in some project right now.

Also, thanks for explaining really well what you are doing, this helps a lot when trying to implement the same stuff.

> **I do not intend to prescribe an exact implementation here or in the future unless someone requests it.** The evolving code is publicly available in the [stringsync repository](https://github.com/jaredjj3/stringsync). I've included some implementations in line with the abstract code.

However, I'm still a little bit lost on how I can implement your code into my OSMD project.
Do you think you can provide any exact implementation, if you have some time ? 


Thanks a lot !

**Comment by bukaznik** [comment]
Hey @jaredjj3! Can we use stringsync.com as an OSMD showcase on https://opensheetmusicdisplay.org/showcases/? Pretty amazing what you put together at stringsync.com! We'd love to show our website audience what OSMD is used for. 

Thanks!

**Comment by jaredjj3** [comment]
@trickart73 thanks for the feedback!

>Do you think you can provide any exact implementation, if you have some time ?

I'm afraid I don't have a great answer right now. An exact implementation is in the [MusicDisplay](https://github.com/jaredjj3/stringsync/tree/master/web/src/lib/MusicDisplay) module of my project. The consuming code lives in different React components/hooks throughout the project ([example](https://github.com/jaredjj3/stringsync/blob/master/web/src/components/NotationPlayer/useMusicDisplayScrollBehaviorEffect.ts)).

Depending on the outcome of this Github issue, I will consider making a plugin/extension as a separate library. However, I strongly prefer something officially supported. From what I understand, the sponsored version already has features like this and I wouldn't want to unravel what the team has already done, especially if it diverges from their long-term goals.

I'd also like to wait to see how my implementation evolves on my end. Maybe it's not as scalable as I think it is.

So ultimately, I'm going to wait. When I do revisit this, I will update this thread.

**Comment by jaredjj3** [comment]
>Hey @jaredjj3! Can we use stringsync.com as an OSMD showcase on https://opensheetmusicdisplay.org/showcases/? Pretty amazing what you put together at stringsync.com! We'd love to show our website audience what OSMD is used for.

@bukaznik, absolutely - I'd be honored! Currently, what's on stringsync.com is actually very stale and is not using OSMD (it's using Vextab/Vexflow). I plan to re-release it in the next month or so with a version that uses OSMD including all the features of the general event system I made. Do you mind if I reach back out to let you know when that's live?

**Comment by sschmidTU** [comment]
> From what I understand, the sponsored version already has features like this and I wouldn't want to unravel what the team has already done, especially if it diverges from their long-term goals.

Yes, that's something that adds extra hours to integrating this into OSMD: there is already an event system in osmd-extended (sponsor OSMD). So at the very least, the two systems would have to be compared and somehow unified.

**Comment by bukaznik** [comment]
> > Hey @jaredjj3! Can we use stringsync.com as an OSMD showcase on https://opensheetmusicdisplay.org/showcases/? Pretty amazing what you put together at stringsync.com! We'd love to show our website audience what OSMD is used for.
> 
> @bukaznik, absolutely - I'd be honored! Currently, what's on stringsync.com is actually very stale and is not using OSMD (it's using Vextab/Vexflow). I plan to re-release it in the next month or so with a version that uses OSMD including all the features of the general event system I made. Do you mind if I reach back out to let you know when that's live?

Yes, this is perfect! Looking forward to it, thanks!

**Comment by jaredjj3** [comment]
bukaznik@, I've done a soft release of stringsync but anyone can access it. I have 1 transcription available at https://stringsync.com/n/NRZF7Cjm. You may have to hard refresh if it's not loading.

Feel free to add stringsync to https://opensheetmusicdisplay.org/showcases/

**Comment by bukaznik** [comment]
@jaredjj3 thanks for the detailed explanation! I was wondering on a more basic level (aka non dev level ;-)) how the syncing is done...do you drop a video file / youtube url + a music xml file into a folder and then the magic happens? Do have or do you plan to make a UI for that?

**Comment by sschmidTU** [comment]
Also, I was wondering how you keep the sheet and audio consistently synced. Do you simply set the start point and run at a certain bpm, so you have to play at a strict tempo, so there's never an offset between the timings? (if you played freely and slowed down at some spot, you'd have to correct for that in the timings)

**Comment by jaredjj3** [comment]
> @jaredjj3 thanks for the detailed explanation! I was wondering on a more basic level (aka non dev level ;-)) how the syncing is done...do you drop a video file / youtube url + a music xml file into a folder and then the magic happens? Do have or do you plan to make a UI for that?
 
@bukaznik  Yeah, I have a simple UI for that, but it's currently gated on the site to only be accessible to me. The user uploads a video and musicXML file. I'm then doing some custom processing so that the video is available in multiple qualities and formats. When that's done, the user has to specify some other info (like the dead time I mentioned). stringsync does all of the heavy lifting to keep the video and the cursor in sync.

The upload/edit UIs look like this:

<img width="1271" alt="Screen Shot 2021-11-23 at 5 33 12 PM" src="https://user-images.githubusercontent.com/19232300/143139699-6811561a-2d38-4ee2-ba69-47e61756b5ee.png">

<img width="1713" alt="Screen Shot 2021-11-23 at 5 17 16 PM" src="https://user-images.githubusercontent.com/19232300/143139619-27850f89-2366-4d40-b007-62f6abb09603.png">

>Also, I was wondering how you keep the sheet and audio consistently synced. Do you simply set the start point and run at a certain bpm, so you have to play at a strict tempo, so there's never an offset between the timings? (if you played freely and slowed down at some spot, you'd have to correct for that in the timings)

@sschmidTU I'm using [videojs](https://videojs.com/) for the video player. I decided that the video player's current time was a reliable source of truth. So almost everything flows downward from there. The entrypoint of this code is [here](https://github.com/stringsync/stringsync/blob/65486853e270904da9757425b0a2f703ee636121/web/src/components/Notation/hooks/useMusicDisplayCursorTimeSync.ts).

However, this is kind of tricky when you want to sync the other way. That is, given a positional event on the music sheet, change the current time in the video player. Instead of changing objects in the music sheet in response to an event, I change the current time such that the resulting music sheet object ends up where I want it. That way, I'm still adhering to the unidirectional flow of data and I can still guarantee that the system reflects the current time. The entrypoint for click and drag interactions are [here](https://github.com/stringsync/stringsync/blob/65486853e270904da9757425b0a2f703ee636121/web/src/components/Notation/hooks/useMusicDisplayPointerInteractions.ts).

**Comment by jaredjj3** [comment]
@bukaznik you might be also interested to know that I'm working on a MusicXML library whose scope is to provide a developer-friendly API around the MusicXML spec. After all, since `opensheetmusicdisplay` is already good at rendering MusicXML, I just have to figure out how to edit it.

**Comment by bukaznik** [comment]
@jaredjj3 one more question...what's your future plans with stringsync?

**Comment by jaredjj3** [comment]
> @jaredjj3 one more question...what's your future plans with stringsync?

One of the main goals is to create a musicXML editor+syncer tailored for guitar that anyone can use. 

A secondary goal would be to create an exercise chain, where users can apply mods to simple exercises and enhance their practice routine. For example, a simple exercise may be “play the major scale”. A mod could be “play the exercise in triplets”. Another mod that can be applied on top of that is “play the exercise for each note in the circle of fifths”. I want to empower users to create their own chains and share them with others.

There’s some other plans like gamifying the application, but that’s just an excuse for me to do a fun problem :)
